### PR TITLE
Switch to ports folder for the platform and http_adapter

### PR DESCRIPTION
**Pull request by RickWinter**
I'm working on better links to the docs and better comments.  Ignore that aspect, which will improve, as you review.
Please focus on the structure and overall architecture.

### CODE REVIEW

File: sdk/ports/_template/az_platform_impl.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+/**
+ * @brief Gets the platform clock in milliseconds.
+ * @remark The moment of time where clock starts is undefined, but if this function is getting
+ * called twice with one second interval, the difference between the values returned should be equal
+ * to 1000.
+ * @return Platform clock in milliseconds.
+ */
+inline AZ_NODISCARD int64_t az_platform_clock_msec() { 
+    #error "go see the docs"
```
Comment (danewalton) [inline sdk/ports/_template/az_platform_impl.h:22]: nit: also in line with previous comment, just to be gentle here and elsewhere in the file
```suggestion
    #error "please see the docs"
```
Should we also have a link to the doc which you are referencing?

File: sdk/ports/_template/az_platform_impl.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ *
+ * @remarks If `*obj` equals to \p expected, its value is being overwritten with \p desired, and no
+ * operation otherwise. Returns `true` if the value was overwritten, `false` otherwise.
+ */
+inline AZ_NODISCARD bool az_platform_atomic_compare_exchange(
+    uintptr_t volatile* obj,
+    uintptr_t expected,
+    uintptr_t desired)
+{
+    #error "go see the docs"
```
Comment (danewalton) [inline sdk/ports/_template/az_platform_impl.h:56]: difference in indentation here?

File: sdk/ports/_template/az_platform_impl.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  return false;
+}
+
+#include <azure/core/_az_cfg_suffix.h>
+
+#endif
```
Comment (danewalton) [inline sdk/ports/_template/az_platform_impl.h:62]: `#endif` tag?
```suggestion
#endif //_az_PLATFORM_IMPL_H
```

File: sdk/ports/posix/az_platform_impl.h [hunk 4]
```diff
@@ -2,25 +2,24 @@
...
-#include <azure/core/az_platform.h>
 
 #include <time.h>
 
 #include <unistd.h>
 
 #include <azure/core/_az_cfg.h>
 
-AZ_NODISCARD int64_t az_platform_clock_msec()
+inline AZ_NODISCARD int64_t az_platform_clock_msec()
```
Comment (danewalton) [inline sdk/ports/posix/az_platform_impl.h:12]: So we are documenting in one place only?

File: sdk/src/azure/core/CMakeLists.txt [hunk 5]
```diff
@@ -63,23 +63,24 @@ add_library (
...
   $<BUILD_INTERFACE:${az_SOURCE_DIR}/sdk/inc>
   $<INSTALL_INTERFACE:include/${TARGET_NAME}>
 )
 
-# include test internal headers
-target_include_directories(${TARGET_NAME}
+target_include_directories (${TARGET_NAME}
   PUBLIC
-    $<BUILD_INTERFACE:${az_SOURCE_DIR}/sdk/tests/core/inc>
+  $<BUILD_INTERFACE:${az_SOURCE_DIR}/sdk/ports/${PORTS_INCUDE_DIR}>
```
Comment (danewalton) [inline sdk/src/azure/core/CMakeLists.txt:77]:
```suggestion
  $<BUILD_INTERFACE:${az_SOURCE_DIR}/sdk/ports/${PORTS_INCLUDE_DIR}>
```

File: sdk/src/azure/core/CMakeLists.txt [hunk 6]
```diff
@@ -63,23 +63,24 @@ add_library (
...
 )
 
-target_link_libraries(${TARGET_NAME}
+# include test internal headers
```
Comment (danewalton) [inline sdk/src/azure/core/CMakeLists.txt:80]: I don't think this comment is valid anymore

File: sdk/ports/_template/az_platform_impl.h [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    return;  
+}
+
+/**
+ * @brief Conditionally exchanges values of two pointers in a
+ * thread-safe manner.
+ *
+ * @param obj A pointer to a pointer that needs to be changed.
+ * @param expected An expected value of a value that \p obj is pointing to, prior to exchange.
+ * @param desired A value to assign to the value that is pointed by \p obj, if its value equals to
```
Comment (danewalton) [inline sdk/ports/_template/az_platform_impl.h:43]: nit:
```suggestion
 * @param desired A value to assign to the value that is pointed by \p obj, if its value is equal to
```

File: sdk/inc/azure/core/az_platform.h [hunk 8]
```diff
@@ -11,42 +11,14 @@
...
- * @return Platform clock in milliseconds.
- */
-AZ_NODISCARD int64_t az_platform_clock_msec();
+#if __has_include("az_platform_impl.h")
+    //Users never directly include az_impl
+    //......
+    #include <az_platform_impl.h>
+#else
+    // Spit out error
+    #error "Error go see the docs"
```
Comment (danewalton) [inline sdk/inc/azure/core/az_platform.h:20]: Here like I commented elsewhere:
```suggestion
    #error "Error please see the docs"
```
with maybe a link to the specific doc

**Comment by danewalton** [comment]
My bad didn't read the top level comment about "ignoring comments and docs links"....spaced out with other stuff. Ignore those from me.

File: sdk/ports/_template/az_platform_impl.h (hunk 1 above)
Comment (CIPop) [reply sdk/ports/_template/az_platform_impl.h:22]: +1 : there should be a reference (readme.md) within the git repo (assume no Internet connectivity)

**Comment by CIPop** [comment]
@gilbertw @JeffreyRichter My preference would be for a single /ports folder under sdk. Having too many top-level folders makes it hard to understand. Also, we don't (shouldn't) support all combinations.
E.g. windows + cURL + OpenSSL should not be a valid combination for security reasons.

The HTTP adapter is, in my mind, something part of ports.

File: sdk/ports/_template/az_platform_impl.h (hunk 1 above)
Comment (danewalton) [reply sdk/ports/_template/az_platform_impl.h:22]: Yea note I made the mistake of tick tacking on these but his top level comment is
> I'm working on better links to the docs and better comments. Ignore that aspect

**Pull request closed without merging**